AnimationCodelab/finished/src/main/java/com/example/android/codelab/animation/ui/home/Home.kt

```diff
@@ -0,0 +1,702 @@
+                )
+            }
+        }
+    }
+}
+
+/**
+ * Shows a message that the edit feature is not available.
+ */
+@OptIn(ExperimentalAnimationApi::class)
+@Composable
+private fun EditMessage(shown: Boolean) {
+    AnimatedVisibility(
+        visible = shown,
+        enter = slideInVertically(
+            // Enters by sliding in from offset -fullHeight to 0.
+            initialOffsetY = { fullHeight -> -fullHeight },
+            animationSpec = tween(durationMillis = 150, easing = LinearOutSlowInEasing)
+        ),
+        exit = slideOutVertically(
+            // Exits by sliding out from offset 0 to -fullHeight.
+            targetOffsetY = { fullHeight -> -fullHeight },
+            animationSpec = tween(durationMillis = 250, easing = FastOutLinearInEasing)
+        )
+    ) {
+        Surface(
+            modifier = Modifier.fillMaxWidth(),
+            color = MaterialTheme.colors.secondary,
+            elevation = 4.dp
+        ) {
+            Text(
+                text = stringResource(R.string.edit_message),
+                modifier = Modifier.padding(16.dp)
+            )
+        }
+    }
+}
+
+/**
+ * Returns whether the lazy list is currently scrolling up.
+ */
+@Composable
+private fun LazyListState.isScrollingUp(): Boolean {
+    var previousIndex by remember(this) { mutableStateOf(firstVisibleItemIndex) }
+    var previousScrollOffset by remember(this) { mutableStateOf(firstVisibleItemScrollOffset) }
+    return remember(this) {
+        derivedStateOf {
+            if (previousIndex != firstVisibleItemIndex) {
+                previousIndex > firstVisibleItemIndex
+            } else {
+                previousScrollOffset >= firstVisibleItemScrollOffset
+            }.also {
+                previousIndex = firstVisibleItemIndex
+                previousScrollOffset = firstVisibleItemScrollOffset
+            }
+        }
+    }.value
+}
+
+/**
+ * Shows the header label.
+ *
+ * @param title The title to be shown.
+ */
+@Composable
+private fun Header(
+    title: String
+) {
+    Text(
+        text = title,
+        style = MaterialTheme.typography.h5
+    )
+}
+
+/**
+ * Shows a row for one topic.
+ *
+ * @param topic The topic title.
+ * @param expanded Whether the row should be shown expanded with the topic body.
+ * @param onClick Called when the row is clicked.
+ */
+@Composable
+private fun TopicRow(topic: String, expanded: Boolean, onClick: () -> Unit) {
+    TopicRowSpacer(visible = expanded)
+    Surface(
+        modifier = Modifier
+            .fillMaxWidth()
+            .clickable(onClick = onClick),
+        elevation = 2.dp
+    ) {
+        Column(
+            modifier = Modifier
+                .fillMaxWidth()
+                .padding(16.dp)
+                .animateContentSize()
+        ) {
+            Row {
+                Icon(
+                    imageVector = Icons.Default.Info,
+                    contentDescription = stringResource(R.string.info)
```

Comment:
Isn't this purely decorative? I think CD should probably be null here.

Reply:
Yes, should be `null`. The surface is clickable so it will merge the semantics properties of all its descendants. "Info" is not a logical part of that merged description.

Reply:
Done